Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp

```diff
@@ -28,39 +28,313 @@
 
 #if ENABLE(WEB_CODECS)
 
+#include "CSSStyleImageValue.h"
+#include "CachedImage.h"
 #include "DOMRectReadOnly.h"
 #include "ExceptionOr.h"
+#include "HTMLCanvasElement.h"
+#include "HTMLImageElement.h"
+#include "HTMLVideoElement.h"
+#include "ImageBitmap.h"
+#include "JSPlaneLayout.h"
+#include "VideoColorSpaceInit.h"
+#include "WebCodecsVideoFrameAlgorithms.h"
 
 namespace WebCore {
 
 WebCodecsVideoFrame::~WebCodecsVideoFrame()
 {
 }
 
-ExceptionOr<Ref<WebCodecsVideoFrame>> WebCodecsVideoFrame::create(CanvasImageSource&&, Init&&)
+static std::optional<Exception> checkImageUsability(const CanvasImageSource& source)
 {
-    return adoptRef(*new WebCodecsVideoFrame);
+    return switchOn(source,
+    [] (const RefPtr<HTMLImageElement>& imageElement) -> std::optional<Exception> {
+        auto* image = imageElement->cachedImage() ? imageElement->cachedImage()->image() : nullptr;
+        if (!image)
+            return Exception { InvalidStateError,  "Image element has no data"_s };
+        if (!image->width() || !image->height())
+            return Exception { InvalidStateError,  "Image element has a bad size"_s };
+        return { };
+    },
+#if ENABLE(CSS_TYPED_OM)
+    [] (const RefPtr<CSSStyleImageValue>& cssImage) -> std::optional<Exception> {
+        auto* image = cssImage->image() ? cssImage->image()->image() : nullptr;
+        if (!image)
+            return Exception { InvalidStateError,  "CSS Image has no data"_s };
+        if (!image->width() || !image->height())
+            return Exception { InvalidStateError,  "CSS Image has a bad size"_s };
+        return { };
+    },
+#endif
+#if ENABLE(VIDEO)
+    [] (const RefPtr<HTMLVideoElement>& video) -> std::optional<Exception> {
+        auto readyState = video->readyState();
+        if (readyState == HTMLMediaElement::HAVE_NOTHING || readyState == HTMLMediaElement::HAVE_METADATA)
```

Comment:
This could be `if (readyState < HTMLMediaElement::HAVE_CURRENT_DATA)`

Do we want to fail if the video doesn't have a video track, or if it has one but the width and/or height are zero?

Reply:
OK, I will add the link to image usability: https://html.spec.whatwg.org/multipage/canvas.html#check-the-usability-of-the-image-argument
If there is no video track or width/height are zero, it should fail later on.